### PR TITLE
Field's impact for ComplexityAnalyzer

### PR DESCRIPTION
**Pull request by SlavaUtesinov**
An ability to specify a particular field's complexity impact by means of `Field<IntGraphType>("id").WithComplexityImpact(123)`, which will be taken into account by `ComplexityAnalyzer`. It is a possible solution of the issues #1254 and #298.

### CODE REVIEW

**Comment by SebastianStehle** [comment]
I would introduce a ComplexityAnalayzerMetadataExtensions  class and provide two methods to set and retrieve the metadata, so that you do not have to know the magic string.

**Comment by Shane32** [comment]
How about a `float` rather than a `decimal`?

File: src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+    public async Task<ComplexityResult> AnalyzeAsync(string query)
+    {
+        var result = await _executer.ExecuteAsync(o =>
+        {
+            o.Query = query;
+            o.RequestServices = _provider;
+        }).ConfigureAwait(false);
+        result.Errors.ShouldBeNull();
+
+        return _analyzer.Analyze(_documentBuilder.Build(query), _config.FieldImpact ?? 2f, _config.MaxRecursionCount, _schema);
```
Comment (sungam3r) [inline src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs:145]:
```suggestion
        return _analyzer.Validate(_documentBuilder.Build(query), _config, _schema);
```

**Review by sungam3r — CHANGES_REQUESTED** [review]

**Comment by Shane32** [comment]
> How about a `float` rather than a `decimal`?

Oh if it's `double` that's okay.  I mean `float` is probably sufficient, but I'm okay with `double`.  I must have misread it earlier.

File: src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs (hunk 1 above)
Comment (SlavaUtesinov) [reply src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs:145]: `Analyze` and `Validate` methods don't have the same return signature, the last returns `void`.

File: src/GraphQL/Validation/Complexity/ComplexityVisitor.cs [hunk 2]
```diff
@@ -29,42 +53,46 @@
             var prevCurrentSubSelectionImpact = context.CurrentSubSelectionImpact;
             var prevCurrentEndNodeImpact = context.CurrentEndNodeImpact;
 
+            var fieldImpact = _visitor?.GetFieldDef()?.GetComplexityImpact() ?? context.AvgImpact;
+            var zeroImpact = fieldImpact == 0;
```
[the remark this answers is not in the document]
Comment (SlavaUtesinov) [reply src/GraphQL/Validation/Complexity/ComplexityVisitor.cs:57]: https://stackoverflow.com/questions/6598179/the-right-way-to-compare-a-system-double-to-0-a-number-int#comment7787303_6598240

File: src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs (hunk 1 above)
Comment (sungam3r) [reply src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs:145]: ok

**Review by sungam3r — APPROVED** [review]

**Comment by sungam3r** [comment]
@Shane32 Merge if OK. In the future, I want to make a few more changes related to the calculation of complexity values.

File: src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    hero { #2
+        id #0
+        name #2
+        ...friendsRoot #132(4/2*66)
+    }
+}
+
+fragment friendsRoot on Hero {
+    friends { #3
+        ...friendsIdName #27(9/2*6)
```
Comment (Shane32) [inline src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs:58]: I don't understand what's going on here, but it doesn't seem correct.  It should just be `3*6` - where 3 is the value of the parent and 6 is the total of the fragment.  Can we fix this while we are making changes here?

File: src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+    [Fact]
+    public async Task MetaData_With_Fragments_ShouldBe_Used()
+    {
+        var result = await _fixture.AnalyzeAsync(@"
+query {
+    hero { #2
+        id #0
+        name #2
+        ...friendsRoot #132(4/2*66)
```
Comment (Shane32) [inline src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs:52]: Similarly, this should be `2*66`, not `4/2*66`.

File: src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs [hunk 5]
```diff
@@ -9,11 +10,11 @@ namespace GraphQL.Validation.Complexity
     public class ComplexityAnalyzer : IComplexityAnalyzer
     {
         /// <inheritdoc/>
-        public void Validate(GraphQLDocument document, ComplexityConfiguration complexityParameters)
+        public void Validate(GraphQLDocument document, ComplexityConfiguration complexityParameters, ISchema? schema = null)
         {
             if (complexityParameters == null)
                 return;
-            var complexityResult = Analyze(document, complexityParameters.FieldImpact ?? 2.0f, complexityParameters.MaxRecursionCount);
+            var complexityResult = Analyze(document, complexityParameters.FieldImpact ?? 2.0f, complexityParameters.MaxRecursionCount, schema);
```
Comment (Shane32) [inline src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs:17]: Why do we have a constant of 2.0f here?  Should it be 2.0f?  It seems like the default impact should be 1.0f.  Either way, perhaps we should convert this to a constant within this class.  For example:

```cs
const float DEFAULT_FIELD_IMPACT = 2.0f
```

File: src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs [hunk 6]
```diff
@@ -9,11 +10,11 @@ namespace GraphQL.Validation.Complexity
     public class ComplexityAnalyzer : IComplexityAnalyzer
     {
         /// <inheritdoc/>
-        public void Validate(GraphQLDocument document, ComplexityConfiguration complexityParameters)
+        public void Validate(GraphQLDocument document, ComplexityConfiguration complexityParameters, ISchema? schema = null)
```
Comment (Shane32) [inline src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs:13]: If this is a breaking change anyway, perhaps we should make `schema` required.

**Review by Shane32 — COMMENTED** [review]
The new code looks good, and I like the added tests.  I have a few questions before we merge, but they could be addressed in future PRs also.

File: src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs (hunk 5 above)
Comment (SlavaUtesinov) [reply src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs:17]: https://github.com/graphql-dotnet/graphql-dotnet/blob/develop/src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs#L16

File: src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs (hunk 6 above)
Comment (SlavaUtesinov) [reply src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs:13]: It still can work without scheme as before

File: src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs (hunk 3 above)
Comment (Shane32) [reply src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs:58]: Ok.  We can address it in a separate issue.  Thanks!

Comment (Shane32) [reply src/GraphQL.Tests/Complexity/ComplexityMetaDataTests.cs:58]: #3164

File: src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs (hunk 5 above)
Comment (sungam3r) [reply src/GraphQL/Validation/Complexity/ComplexityAnalyzer.cs:17]: @Shane32 Another observation from me - look at `GetImpactFromArgs`. If we have `id` argument then method returns 1 so `CurrentEndNodeImpact` will be less 2 times. Why so? I don't know.

**Comment by sungam3r** [comment]
> My intention was to implement an ability to specify a specific field's complexity not to rewrite core calculation logic of complexity.

Agree. Complexity is not an easy area to design.

**Pull request merged**